### ISSUE TITLE
Add network cache functionality to video_player

### ISSUE THREAD

**Issue by nagoya0** [flutter/flutter#28094]
## Steps to Reproduce
I built this sample:

https://github.com/flutter/plugins/tree/master/packages/video_player/example
(v0.10.0+2)

and switched tab from side to side. Downloaded videos are disappear and started redownloading every time.

## Flutter Doctor

[√] Flutter (Channel beta, v0.11.10, on Microsoft Windows [Version 10.0.17763.316], locale ja-JP)
    • Flutter version 0.11.10 at C:\flutter
    • Framework revision c27c4a265e (3 months ago), 2018-11-26 17:07:24 -0500
    • Engine revision eebc6a5895
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at C:\Users\XXXXXXXX\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 33.0.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[√] Connected device (1 available)
    • Pixel 3 • XXXXXXXXX • android-arm64 • Android 9 (API 28)

• No issues found!
```

**Comment by Agraphie** [comment]
Is there any intermediate solution for this problem? Otherwise any app which uses videos will drain your data volume quite fast.

**Comment by ohjs88** [comment]
I cached the data by adding some code as shown below.
(I modified video_player 0.10.1+6.)


**VideoPlayerPlugin.java**

Added code
`
private static SimpleCache sDownloadCache;

    class CacheDataSourceFactory implements DataSource.Factory {
      private final Context context;
      private final DefaultDataSourceFactory defaultDatasourceFactory;
      private final long maxFileSize, maxCacheSize;

      CacheDataSourceFactory(Context context, long maxCacheSize, long maxFileSize) {
        super();
        this.context = context;
        this.maxCacheSize = maxCacheSize;
        this.maxFileSize = maxFileSize;
        //String userAgent = Util.getUserAgent(context, context.getString("ExoPlayer"));
        DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter();
        defaultDatasourceFactory = new DefaultDataSourceFactory(this.context, bandwidthMeter,
            new DefaultHttpDataSourceFactory("ExoPlayer", bandwidthMeter,DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS, DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS,true));
      }

      @Override
      public DataSource createDataSource() {
        LeastRecentlyUsedCacheEvictor evictor = new LeastRecentlyUsedCacheEvictor(maxCacheSize);
        
        if (sDownloadCache == null) {
            sDownloadCache = new SimpleCache(context.getCacheDir(),evictor);
        }

        return new CacheDataSource(sDownloadCache, defaultDatasourceFactory.createDataSource(), new FileDataSource(),
            new CacheDataSink(sDownloadCache, maxFileSize),
            CacheDataSource.FLAG_BLOCK_ON_CACHE | CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR, null);
      }
    }
`

Modified code
`
DataSource.Factory dataSourceFactory;
      if (isFileOrAsset(uri)) {
        dataSourceFactory = new DefaultDataSourceFactory(context, "ExoPlayer");
      } else {
        // dataSourceFactory = new DefaultHttpDataSourceFactory("ExoPlayer", null,
        //     DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS, DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS,
        //     true);
        dataSourceFactory = new CacheDataSourceFactory(context, 100 * 1024 * 1024, 5 * 1024 * 1024);
      }
`

**Comment by lifenautjoe** [comment]
Any other way so far that doesn't involve monkey-patching the library?

**Comment by lifenautjoe** [comment]
Here's a plugin for the iOS AV player that might help. https://github.com/vitoziv/VIMediaCache

**Comment by lukepighetti** [comment]
I need video caching as well, iOS and Android with video_player, does anyone have time to put this together?

**Comment by lukepighetti** [comment]
@ohjs88 Can you make a PR against video_player for your cached android code? Do you happen to have iOS code as well?

**Comment by lukepighetti** [comment]
Is it possible to get the file out of `video_player` and save it to disk with Dart code? Not sure if that's obscene or not.

**Comment by lukepighetti** [comment]
Has anyone tried caching the loaded VideoPlayerControllers in Dart?

**Comment by lukepighetti** [comment]
I tried that out and it does work but you end up with some strange behavior. The app runs out of memory real quick.

**Comment by lifenautjoe** [comment]
For what's worth, @999eagle got caching working on both iOS and android on a custom fork of the plugins repository video_player. We hope to open a PR with this back into the plugins repo at some point.

```
  video_player:
    git:
      url: https://github.com/999eagle/plugins.git
      ref: feature/caching
      path: packages/video_player
```

Limit is set to 200mbs


Here are the commits with the magic.

https://github.com/999eagle/plugins/commits/feature/caching

Thanks to her work, we've now got video playing in our app :-)

**Comment by lukepighetti** [comment]
@lifenautjoe Thanks for the heads up, is there any reason why there hasn't been a PR back to flutter/plugins ?

**Comment by lifenautjoe** [comment]
We're busy with post release activities 😬 , we just got this working a couple days back.

**Comment by lukepighetti** [comment]
Much appreciate for the updates, I'm going to watch that repo for movement.

**Comment by lukepighetti** [comment]
@lifenautjoe Do you guys have a commit cherry picked? This branch won't load a `.mov` file that otherwise works with `video_player: ^0.10.2+1`

**Comment by lifenautjoe** [comment]
Ah I just recalled there’s an issue to solve still on iOS for playing local
videos. 😔

We had a merge conflict in regards to how the source of the video is
handled.

‘’’
On Android they changed the function to check the video source. Previously
they handled every full URI (including file:///) as network, now they check
whether the scheme is 'http' or 'https'
‘’’

On iOS we didn’t see any change on this, so we believe it’s trying to load
the local videos as /path/video.mp4 instead of file://path/video.mp4

Meaning videos from network will play, but local videos won’t...

We’ll try to figure this one out in the upcoming days.

We’ve opened an issue on the flutter repo as well but on my phone so can’t
post it now.

**Comment by lukepighetti** [comment]
@lifenautjoe Hi in this case it is a network video with the URL https://firebasestorage.googleapis.com/v0/b/skatex-8d802.appspot.com/o/test%2Ftest.mov?alt=media&token=a59686ca-09cc-4be4-ba70-5069b3c42da7 I might be misunderstanding the issue, any more explanation? I'll keep poking thru the source.

**Comment by lukepighetti** [comment]
It appears to be unable to handle query params? this works fine

http://techslides.com/demos/sample-videos/small.mp4

**Comment by lukepighetti** [comment]
Another case....

This works on the pub.dev version of `video_player` but fails on the `feature/caching` version

https://firebasestorage.googleapis.com/v0/b/skatex-8d802.appspot.com/o/test%2Ftest-Broadband%20Low.mp4?alt=media&token=00f64abd-c552-4d7b-a964-72a22811722f

However, it will not play on either as a percent encoded url
https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fskatex-8d802.appspot.com%2Fo%2Ftest%252Ftest-Broadband%2520Low.mp4%3Falt%3Dmedia%26token%3D00f64abd-c552-4d7b-a964-72a22811722f

**Comment by lukepighetti** [comment]
I've been fooling around with this for a few hours now, tried the following and still getting the `-1002 "unsupported url"` error

```
        NSString *escapedURL = [uriArg stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLQueryAllowedCharacterSet];
```

```
        NSString *escapedURL = [uriArg stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
```

```
        NSString *escapedURL = [uriArg stringByAddingPercentEscapesUsingEncoding:NSUTF16StringEncoding];
```

**Comment by lukepighetti** [comment]
Pull request: https://github.com/999eagle/plugins/pull/1/commits/ed5e865aeae128c2ebf4e500ff4dfb0c79133163

**Comment by lukepighetti** [comment]
@lifenautjoe @999eagle just got to it and `VideoPlayerController.file()` does not work on iOS on 999eagle/plugins/video_player @ feature/caching. Any insights?

**Comment by lukepighetti** [comment]
I believe this makes `feature/caching` ready for PR on iOS. I haven't tried Android yet

https://github.com/999eagle/plugins/pull/2

**Comment by recastrodiaz** [comment]
I have an alternative video networking caching implementation if anyone is interested in using it:

https://github.com/flutter/plugins/compare/master...recastrodiaz:develop2

It has other features as well like:

- audio focus
- clipping videos
- playback speed feature
- fixes some videos showing a black screen on iOS
- more recent ExoPlayer

**Comment by lukepighetti** [comment]
I’m going to check it out because you’ve added some features I need myself. Will you be making a PR against video_player? Is it cross platform?

**Comment by recastrodiaz** [comment]
It is indeed cross platform (Android + iOS):

There are separate PRs for most of the features except the caching of videos and clipping:

- ExoPlayer upgrade: https://github.com/flutter/plugins/pull/1868
- Black screen: https://github.com/flutter/plugins/pull/1307
- Playback speed: https://github.com/flutter/plugins/pull/1400

I'm not planning on doing PRs for the other features for now as these PRs will most likely stay unmerged.

**Comment by lifenautjoe** [comment]
Hi Rodrigo ! (@recastrodiaz ) How does the caching configuration work on your solution? 🤔

**Comment by recastrodiaz** [comment]
Hi @lifenautjoe The cache settings are hard-coded:
- Android: https://github.com/flutter/plugins/compare/master...recastrodiaz:develop2#diff-d0f10ef0e25e308a35f99febd07efc88R19 and https://github.com/flutter/plugins/compare/master...recastrodiaz:develop2#diff-f88f2c8adb7a0c7064d8b69b1f14b0edR348
- iOS (uses default cache `VIResourceLoaderManager`settings): https://github.com/flutter/plugins/compare/master...recastrodiaz:develop2#diff-e534c460662d39cc715c4eed77a31c3fR32

**Comment by lifenautjoe** [comment]
Thanks for the quick reply!

With the latest flutter/plugins master and flutter master version, our solution is not working.

`videoPlayerController.initialize()` fails with errors of the following type

```
2019-10-07 19:23:44.422820+0200 Runner[4520:1777202] Task <2916A884-BCD7-4AC2-8DF8-50297E8F5884>.<1> finished with error [-1002] Error Domain=NSURLErrorDomain Code=-1002 "unsupported URL" UserInfo={NSLocalizedDescription=unsupported URL, NSErrorFailingURLStringKey=https%3A%2F%2Fassets%2Dapi%2Dopenbook%2Dsocial%2Es3%2Daccelerate%2Eamazonaws%2Ecom%2Fmedia%2Fpublic%2Fformats%2Fmp4%5Fsd%2Fposts%2F3b8683f9%2Dbda1%2D422a%2D9f61%2D7e5a126f7b90%2F70678389%2D5265%2D41db%2Da92e%2D1a8a8996a318%2Emp4, NSErrorFailingURLKey=https%3A%2F%2Fassets%2Dapi%2Dopenbook%2Dsocial%2Es3%2Daccelerate%2Eamazonaws%2Ecom%2Fmedia%2Fpublic%2Fformats%2Fmp4%5Fsd%2Fposts%2F3b8683f9%2Dbda1%2D422a%2D9f61%2D7e5a126f7b90%2F70678389%2D5265%2D41db%2Da92e%2D1a8a8996a318%2Emp4, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <2916A884-BCD7-4AC2-8DF8-50297E8F5884>.<1>"
), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <2916A884-BCD7-4AC2-8DF8-50297E8F5884>.<1>, NSUnderlyingError=0x283a4fba0 {Error Domain=kCFErrorDomainCFNetwork Code=-1002 "(null)"}}
```

I wanted to try with your solution to see if it was the same case. Some 50% of the time videos fail too.

In the logs I see

```
2019-10-17 15:57:10.678838+0200 Runner[1407:370002] VIDEO__ 0.000000, 0.000000, 480.000000, 852.000000, 0
2019-10-17 15:57:11.077348+0200 Runner[1407:369906] Task <33027A16-19C4-477E-B4D2-0A403D1EBF6B>.<1> HTTP load failed, 0/0 bytes (error code: -999 [1:89])
2019-10-17 15:57:11.078681+0200 Runner[1407:369906] Connection 4: unable to determine interface type without an established connection
2019-10-17 15:57:11.078709+0200 Runner[1407:369906] Connection 4: unable to determine fallback status without a connection
```

Had you noticed this too?

CC @lukepighetti @999eagle

**Comment by recastrodiaz** [comment]
I noticed my cached videos don't always open the first time around. On the second try they always open and play correctly. It may be the same issue you are experiencing but I haven't looked closely at the issue.

**Comment by lifenautjoe** [comment]
Sounds about right! Unfortunately I'm limited when it comes to iOS. Let me know if there's another way I can help resolve this.

**Comment by lifenautjoe** [comment]
Another thing noticed, on the current master of flutter/plugins, there's no loading issue. So I'm assuming it's on the caching part.

**Comment by Nisthar** [comment]
any updates on this?

**Comment by blbhargav** [comment]
> any updates on this?
https://pub.dev/packages/flutter_cache_manager
Using this plugin with video player. Together works perfectly.

**Comment by sanekyy** [comment]
Hello everyone. I and @otopba created [new PR](https://github.com/flutter/plugins/pull/2429) which fix this previous PR from @999eagle to speed up the process of merging of this feature.

We hope for your support

**Comment by recastrodiaz** [comment]
I had a closer look today at why http videos don't always play the first time when using the cache. For some reason we attempt to use the video in the `AVPlayer` when it is not yet ready. I've worked around this by simply loading it again in case of such an error:

```LOG
Could not load video tracks: 3 - Error Domain=AVFoundationErrorDomain
Code=-11829 "Cannot Open" UserInfo={NSUnderlyingError=0x28117e190 {Error
Domain=NSOSStatusErrorDomain Code=-12848 "(null)"}, NSLocalizedFailureReason=This media may be
damaged., NSURL=__VIMediaCache___:https://example.com/video.mp4, NSLocalizedDescription=Cannot
Open}

```

If someone has a better approach to fix this feel free to reach out.

The commit in question is: https://github.com/recastrodiaz/plugins/commit/9bc2ac1d07d1f159690395d25e01861abd5aa685

That branch has many other features highlighted in https://github.com/flutter/flutter/issues/28094#issuecomment-538461443

**Comment by giiyms** [comment]
> > any updates on this?
> > https://pub.dev/packages/flutter_cache_manager
> > Using this plugin with video player. Together works perfectly.

@blbhargav  Did you have an example you can provide?

**Comment by harshapulikollu** [comment]
> > > any updates on this?
> > > https://pub.dev/packages/flutter_cache_manager
> > > Using this plugin with video player. Together works perfectly.
> 
> @blbhargav Did you have an example you can provide?

Here is how I'm caching video with **flutter_cache_manager** and **chewie**. Hope it will be helpful.

```
class FullScreenImage extends StatefulWidget {
  final DocumentSnapshot document;

  MsgFullScreenImage({this.document});
  @override
  _MsgFullScreenImageState createState() => _MsgFullScreenImageState();
}

class _MsgFullScreenImageState extends State<MsgFullScreenImage> {

//widget.document['mediaUrl'] is the url of the video

  ChewieController chewieController;
  VideoPlayerController videoPlayerController;
  File file;
  bool fetchVideoFromOnline = true;

  @override
  void initState() {
    initPlatformState();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Container(
       child: getMediaWidget())
    );
  }

  getMediaWidget() {
      videoPlayerController = fetchVideoFromOnline
          ? VideoPlayerController.network(widget.document['mediaUrl']) //url of video
          : VideoPlayerController.file(file); // cached video file

      chewieController = ChewieController(
        placeholder: BlurHash(
          hash: widget.document['placeholder'],
        ),
        videoPlayerController: videoPlayerController,
        aspectRatio: 3 / 2,
        autoPlay: true,
        looping: true,
      );
      return Chewie(
        controller: chewieController,
      );

  }

  void initPlatformState() async {

    FileInfo fileInfo = await DefaultCacheManager().getFileFromCache(widget.document['mediaUrl']);//url of video

    if (fileInfo.file == null) {
      print('cache ln: caching now ');

      setState(() {
        fetchVideoFromOnline = true;
      });

      file = await DefaultCacheManager().getSingleFile(widget.document['mediaUrl']); //here we provide the url of video to cache.
    } else {
      print('cache ln: ${fileInfo.validTill}');
      setState(() {
        fetchVideoFromOnline = false;
        file = fileInfo.file;
      });
    }
  }

@override
  void dispose() {
      videoPlayerController.dispose();
      chewieController.dispose();
    super.dispose();
  }
}
```

**Comment by Yodart** [comment]
@harshapulikollu  Your solution doesn't seem to work on iOS, or I'm a wrong here?

**Comment by harshapulikollu** [comment]
> @harshapulikollu Your solution doesn't seem to work on iOS, or am i wrong here?

Hi @Yodart , I have tested this code on iOS simulator after your comment.
Firstly with network connected(In order to cache the file). 
Later with turning off wifi of Macbook to stop network connectivity to iOS simulator and as expected the audio was able to play and **video won't play in iOS simulators** according to this flutter issue response https://github.com/flutter/flutter/issues/14647#issuecomment-365038838
If you have tested it on physical iOS device or any other playback related issue or if you had found any workarrounds, please update me.

**Comment by myomyintaung1411** [comment]
> > > > any updates on this?
> > > > https://pub.dev/packages/flutter_cache_manager
> > > > Using this plugin with video player. Together works perfectly.
> > 
> > 
> > @blbhargav Did you have an example you can provide?
> 
> Here is how I'm caching video with **flutter_cache_manager** and **chewie**. Hope it will be helpful.
> 
> ```
> class FullScreenImage extends StatefulWidget {
>   final DocumentSnapshot document;
> 
>   MsgFullScreenImage({this.document});
>   @override
>   _MsgFullScreenImageState createState() => _MsgFullScreenImageState();
> }
> 
> class _MsgFullScreenImageState extends State<MsgFullScreenImage> {
> 
> //widget.document['mediaUrl'] is the url of the video
> 
>   ChewieController chewieController;
>   VideoPlayerController videoPlayerController;
>   File file;
>   bool fetchVideoFromOnline = true;
> 
>   @override
>   void initState() {
>     initPlatformState();
>     super.initState();
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       backgroundColor: Colors.black,
>       body: Container(
>        child: getMediaWidget())
>     );
>   }
> 
>   getMediaWidget() {
>       videoPlayerController = fetchVideoFromOnline
>           ? VideoPlayerController.network(widget.document['mediaUrl']) //url of video
>           : VideoPlayerController.file(file); // cached video file
> 
>       chewieController = ChewieController(
>         placeholder: BlurHash(
>           hash: widget.document['placeholder'],
>         ),
>         videoPlayerController: videoPlayerController,
>         aspectRatio: 3 / 2,
>         autoPlay: true,
>         looping: true,
>       );
>       return Chewie(
>         controller: chewieController,
>       );
> 
>   }
> 
>   void initPlatformState() async {
> 
>     FileInfo fileInfo = await DefaultCacheManager().getFileFromCache(widget.document['mediaUrl']);//url of video
> 
>     if (fileInfo.file == null) {
>       print('cache ln: caching now ');
> 
>       setState(() {
>         fetchVideoFromOnline = true;
>       });
> 
>       file = await DefaultCacheManager().getSingleFile(widget.document['mediaUrl']); //here we provide the url of video to cache.
>     } else {
>       print('cache ln: ${fileInfo.validTill}');
>       setState(() {
>         fetchVideoFromOnline = false;
>         file = fileInfo.file;
>       });
>     }
>   }
> 
> @override
>   void dispose() {
>       videoPlayerController.dispose();
>       chewieController.dispose();
>     super.dispose();
>   }
> }
> ```

hello bro how to cache list of video ?

**Comment by saurabhrayakwar** [comment]
https://pub.dev/packages/cached_video_player#-readme-tab-

**Comment by jpiabrantes** [comment]
hey @harshapulikollu 

When the cache is empty are you downloading the same video twice by doing:
```dart 
file = await DefaultCacheManager().getSingleFile(widget.document['mediaUrl']); //here we provide the url of video to cache.
```
and
```dart 
VideoPlayerController.network(widget.document['mediaUrl'])
```

Or is there something smart happening under the hood? If not, would it be possible to get access to the data downloaded by the VideoPlayerController and copy it to the cache?

**Comment by Raikan10** [comment]
Hey guys,
I'm trying to figure this problem out too, I cam across this dependency called [cached_video_player](https://pub.dev/packages/cached_video_player). I'm not sure how exactly it works but maybe it will help for some of you.

P.S: It doesn't work with chewie you will have to enable your own controls.

**Comment by harshapulikollu** [comment]
> hey @harshapulikollu
> 
> When the cache is empty are you downloading the same video twice by doing:
> 
> ```dart
> file = await DefaultCacheManager().getSingleFile(widget.document['mediaUrl']); //here we provide the url of video to cache.
> ```
> 
> and
> 
> ```dart
> VideoPlayerController.network(widget.document['mediaUrl'])
> ```
> 
> Or is there something smart happening under the hood? If not, would it be possible to get access to the data downloaded by the VideoPlayerController and copy it to the cache?

Hi @jpiabrantes, yes as you said it downloads twice when cache is empty(As per my knowledge).
I couldn't spend much time on it. Feel free to update us if you can minimize it to once when cache is empty.

**Comment by JonasJW** [comment]
Hi @harshapulikollu 

To my knowledge, with your code above you download the whole video before being able to play it. While `VideoPlayer.network(...)` uses streaming, so when you want to play a 2 min. video it takes a lot longer to start the video with you caching method while with `VideoPlayer.network(...)` you need to load only the first few seconds before being able to start it. Or am I missing something here?

**Comment by afnx** [comment]
Maybe this could be helpful until video caching implemented...

```dart
import 'dart:io';
import 'package:flutter_cache_manager/flutter_cache_manager.dart';
import 'package:video_player/video_player.dart';
import 'package:flutter/material.dart';

class VideoDisplay extends StatefulWidget {
  final String videoUrl;

  VideoDisplay({this.videoUrl});

  @override
  _VideoDisplayState createState() => _VideoDisplayState();
}

class _VideoDisplayState extends State<VideoDisplay> {
  VideoPlayerController _controller;
  bool _isVideoLoaded = false;

  @override
  void initState() {
    if (!_isVideoLoaded) {
      setState(() => _isVideoLoaded = true);
      _downloadAndCacheVideo().then((file) {
       if (file != null) {
          _controller = VideoPlayerController.file(file)
            ..initialize().then((_) {
              setState(() {});
              _controller.play();
            });
          });
       }
    }

    super.initState();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  Future<File> _downloadAndCacheVideo() async {
    try {
      final cacheManager = DefaultCacheManager();

      FileInfo fileInfo;

      fileInfo = await cacheManager.getFileFromCache(widget.videoUrl); // Get video from cache first

      if (fileInfo?.file == null) {
        fileInfo = await cacheManager.downloadFile(widget.videoUrl); // Download video if not cached yet
      }

      return fileInfo?.file;
    } catch (e) {
      throw (e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: _controller?.value?.initialized == true
          ? VideoPlayer(_controller)
          : Center(
              child: CircularProgressIndicator(),
            ),
    );
  }
}
```

**Comment by wanjm** [comment]
@afnx doesn't support hls?

**Comment by Gicminos** [comment]
Long time is passing for the PR to get merged, and many people are waiting or looking for temporary workarounds. I wanted to both make a summary on the messages of the issue, and both wanted to ask what the other users ended up doing. 
Between:

- Using the [cached_video_player plugin](https://pub.dev/packages/cached_video_player)
- Using the [fork currently under PR](https://github.com/sanekyy/plugins/tree/caching)
- Using the default plugin with [flutter_cache_manager](https://pub.dev/packages/flutter_cache_manager)
- Using many custom code snippets that have been published in the issue or somewhere else

What is the best solution for now? What you people ended up doing? Your feedback would be extrimely helpful for me and possibly to everyone ending up in this issue. Thanks

**Comment by JonasJW** [comment]
@Gicminos I'm using a PR of the fork. The disadvantage is I can't use the latest version of the video player with the mixinOption, other than that, it works fine so far.

**Comment by lukepighetti** [comment]
I'm sorry to say that I am super busy right now and won't be able to take over this PR if @sanekyy . Is there anyone else willing to pick up where he left off?

**Comment by nagoya0** [comment]
It's been two years since this issue was created.
Compared to before, the number of platforms that need to be supported has increased, I think it has become more difficult to implement, but realistically, I would want to avoid using video_player unless this issue is resolved...

**Comment by omidraha** [comment]
I have same issue and currently use [flutter_cache_manager]( https://pub.dev/packages/flutter_cache_manager) ,
Here is my sample code:

Bloc:
```dart
      yield PlayerIsLoading();
      String url = 'sample video url';
      try {
        file = await userRepository.getFileAPI(url);
      } on ServerConnectionFailed catch (e) {
        yield ServerIsNotAvailable(
            networkMessage: e.message,
        );
        return;
      }

      VideoPlayerController videoPlayerController =
          VideoPlayerController.file(file);

      yield PlayerLoaded(
        videoPlayerController: videoPlayerController,
      );
    }
```

userRepository:
```dart

  Future<File> getFileAPI(String url) async {
    File file;
    try {
      file = await DefaultCacheManager().getSingleFile(url).timeout(
            Duration(seconds: SocketTimeOut),
          );
    } on HttpException catch (e) {
      throw ServerConnectionFailed(
        message: ServerConnectionFailed.defaultMessage,
      );
    } on TimeoutException catch (e) {
      throw ServerConnectionFailed(
        message: ServerConnectionFailed.defaultMessage,
      );
    } on SocketException catch (e) {
      throw ServerConnectionFailed(
        message: ServerConnectionFailed.defaultMessage,
      );
    }
    return file;
  }
```

Note:
The `CacheManager` package  by default download [old](https://github.com/Baseflow/flutter_cache_manager/blob/3bd49b866d9b5e6ce09d2283f0ef97488be97e1d/flutter_cache_manager/lib/src/cache_manager.dart#L70) file [again](https://github.com/Baseflow/flutter_cache_manager/blob/3bd49b866d9b5e6ce09d2283f0ef97488be97e1d/flutter_cache_manager/lib/src/cache_manager.dart#L206) (after 30 day), this is not suitable for videos file you know there is no change in lifetime of them.

**Comment by b3nni97** [comment]
Is anyone currently working on a solution? Caching is a critical feature for many apps

**Comment by JonasJW** [comment]
@b3nni97 check out the better_player plugin, it solved my video caching problems

**Comment by amalnaami** [comment]
> @b3nni97 check out the better_player plugin, it solved my video caching problems

It doesn't solve the problem 
there is any new suggestion to cache video from network ??

**Comment by b3nni97** [comment]
I think the only thing that helps is to wait until many people become aware of the problem so that something happens. 
The current video player is also not very efficient when you want to display multiple videos in a ListView. I haven't found an approach that works yet. (there are lags etc)

**Comment by sanekyy** [comment]
@b3nni97 
Hello. For this case you can try to use my old PR with reusable video player:)
https://github.com/flutter/plugins/pull/2406

**Comment by b3nni97** [comment]
@sanekyy 
Thanks for the tip. The problem is that such pull requests should be merged. I see the problem is due to an unsigned CLA. I would be very grateful if you could catch up on this, because this is a critical feature (for me at least). 

I think that such basic features should be supported by default in the video player.

**Comment by amalnaami** [comment]
I wondered if we can stream video Url that come from back using Flutter code ? 

> @sanekyy
> Thanks for the tip. The problem is that such pull requests should be merged. I see the problem is due to an unsigned CLA. I would be very grateful if you could catch up on this, because this is a critical feature (for me at least).
> 
> I think that such basic features should be supported by default in the video player.

**Comment by kamlesh1687** [comment]
> I think the only thing that helps is to wait until many people become aware of the problem so that something happens. 
> The current video player is also not very efficient when you want to display multiple videos in a ListView. I haven't found an approach that works yet. (there are lags etc)

You can use Flick player ,they have implemented their own way for playing videos in feed

**Comment by Leffe108** [comment]
@Gicminos I use flutter_cache_manager with video_player. In my case I am often able to know ahead of time which video to play ahead of time so I can trigger download to cache. Thus in many cases I avoid duplicate downloads and this solution works quite well. It also helps in my case that the videos are rather short.

To give a bit more context the app is a quiz app and when user start to answer one question, I generate the next one and load video clips for the next one while user is responding to the current question. Usually the user is slower. :-)

**Comment by mavericksunny** [comment]
https://github.com/mavericksunny/plugins/commits/caching 

@sanekyy Ive forked your repository and updated it with the latest video player master repo. Took me a while to resolve the conflicts, but it seems to work now. 
For anyone who wants to use it, you can use this as a dependency:

video_player:
   git:
    url: git@github.com:mavericksunny/plugins.git
    ref: caching
    path: packages/video_player/video_player

**Comment by lukepighetti** [comment]
@mavericksunny open a PR if you have it working and up to date with main branch, maybe we can get this merged once and for all

**Comment by sanekyy** [comment]
@mavericksunny 
Great!
Link PR here if you'll open

**Comment by aytunch** [comment]
I hope it is working as intended and gets approved. This can be a huge improvement for video driven apps.

**Comment by mavericksunny** [comment]
@sanekyy Ive opened it against your repo: https://github.com/sanekyy/plugins/pull/5
Also can you merge the latest master from flutter plugins? To avoid seeing the complete diff.

https://github.com/sanekyy/plugins/pull/5/commits/677d928d1a745e9bdf4257163daa2c52a5374fef

**Comment by marcusrohden** [comment]
> @sanekyy Ive opened it against your repo: [sanekyy/plugins#5](https://github.com/sanekyy/plugins/pull/5) Also can you merge the latest master from flutter plugins? To avoid seeing the complete diff.
> 
> [sanekyy/plugins@677d928](https://github.com/sanekyy/plugins/commit/677d928d1a745e9bdf4257163daa2c52a5374fef)

hey guys, any updates on this? the caching feature would be amazing to have.
thx!

**Comment by ciriousjoker** [comment]
We ended up using `DefaultCacheManager()` to download the video before displaying it.
This does work, but the drawback is obvious. The whole video is downloaded before displaying a single frame, which is terrible UX. The alternative would have been to stream the video first and cache it in the background, which would download the whole thing twice.

Please add caching to video_player so we can remove our custom caching logic and stream the files directly without caching first or downloading it twice.

**Comment by aytunch** [comment]
In our app we have a video feed and we are preloading 2-3 videos ahead of time which makes the bandwidth usage even worse with the current video_player implementation. If there were caching support, the feed would have been much more memory, performance and bandwidth effective which would result in better user feedback..

Please up-vote the issue so this takes into consideration with a higher priority.

**Comment by b3nni97** [comment]
I also don't understand why nothing happens here. Caching is essential especially in the mobile area where bandwidth and memory are relatively scarce.

**Comment by lukepighetti** [comment]
@TahaTesser Sorry to ping. Is this the type of feature that should have a doc written up first? Would appreciate some guidance / prior buy in before investing time into this.

**Comment by TahaTesser** [comment]
cc: @stuartmorgan 
Could you please provide some guidance https://github.com/flutter/flutter/issues/28094#issuecomment-1030951903

**Comment by stuartmorgan** [comment]
> Is this the type of feature that should have a doc written up first?

It certainly never hurts to have a design document for a non-trivial PR.

> Don't want a repeat of the last time the community attempted to land this

My understanding is that the primary issue last time was that the PR went unreviewed for a very long time. That should no longer happen, as flutter/plugins PRs now follow the same triage policy as the other Flutter repositories, meaning we are regularly checking in with all open PRs.

**Comment by lukepighetti** [comment]
Is there anyone in particular who should be pinged when a design doc is ready for this type of feature?

**Comment by drose1db** [comment]
Hey guys, are you planning to make this feature soon? We've been waiting for this for a long time.

**Comment by victor-tinoco** [comment]
I'm also needing this feature ASAP. Any update?

**Comment by iosephmagno** [comment]
Any news on this? Will be a merge coming soon?

**Comment by khater277** [comment]
.file not work correctly it gives me that error in initialize  [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: PlatformException(NullPointerException, java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference, null, null)

**Comment by khater277** [comment]
@mavericksunny .file not work correctly it gives me that error in initialize [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: PlatformException(NullPointerException, java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference, null, null)

**Comment by Chetan-3110** [comment]
For handling cache, I am currently maintaining a list of controllers and disposing them depending on the position in the list. The problem here is that the app starts consuming a lot of memory (since my videos are of length >= 1 minute) which in turn crashes the app multiple times. 
So, definitely not a good approach but could be a possible workaround in some cases.

**Comment by Helocominc** [comment]
Any solution to the caching problem? I have been avoiding using video player for anything due to its inability to cache videos for over 2 years now. Is there any follow up or work around yet?

**Comment by fzyzcjy** [comment]
Hi, is there any updates :/ It has been more than 3.5 years and IMHO it is very commonly used as long as one needs to play a video

**Comment by marcusrohden** [comment]
> Any solution to the caching problem? I have been avoiding using video player for anything due to its inability to cache videos for over 2 years now. Is there any follow up or work around yet?

Unfortunately, this issue is marked as P4 which stands for: they may never fix it.

more details:
https://github.com/flutter/flutter/wiki/Issue-hygiene#when-will-my-bug-be-fixed

**Comment by JonasJW** [comment]
> Would be great to see if someone has any workarounds

My workaround is using the `better_player` Package which supports caching out of the box and support all other `video_player` functionalities.

**Comment by farzasim1** [comment]
> For what's worth, @999eagle got caching working on both iOS and android on a custom fork of the plugins repository video_player. We hope to open a PR with this back into the plugins repo at some point.
> 
> ```
>   video_player:
>     git:
>       url: https://github.com/999eagle/plugins.git
>       ref: feature/caching
>       path: packages/video_player
> ```
> 
> Limit is set to 200mbs
> 
> Here are the commits with the magic.
> 
> https://github.com/999eagle/plugins/commits/feature/caching
> 
> Thanks to her work, we've now got video playing in our app :-)

It is working but how can I clear the cache if I need to?

**Comment by otopba** [comment]
Happy 2023 everyone. Another year without video caching has passed

**Comment by delfme** [comment]
Yep remote video streaming without caching is not viable solution for social media / messaging apps.

We use better_player which provides Cache feature and we also added to it preCache. If preCache is enabled, the controller will cache video for only X ms (we set X = 350ms). This is enough to let video playback starts with no delay and at same time save cost of bandwidth. Lets consider a social media app caching tens of videos at once, that would be pricely, inefficient and unperformant. So preCache helps much here, we preCache them all and cache only the ones which get played.

Wish flutter team gets to enable both Cache/preCache this year. Although we have a solution, we think video playback should not rely on external plugins.

**Comment by MElochukwu** [comment]
> Yep remote video streaming without caching is not viable solution for social media / messaging apps.
> 
> We use better_player which provides Cache feature and we also added to it preCache. If preCache is enabled, the controller will cache video for only X ms (we set X = 350ms). This is enough to let video playback starts with no delay and at same time save cost of bandwidth. Lets consider a social media app caching tens of videos at once, that would be pricely, inefficient and unperformant. So preCache helps much here, we preCache them all and cache only the ones which get played.
> 
> Wish flutter team gets to enable both Cache/preCache this year. Although we have a solution, we think video playback should not rely on external plugins.

This is a good suggestion and the preCache thing will be great. How can we make this come to live?

**Comment by riveraj33** [comment]
> Yep remote video streaming without caching is not viable solution for social media / messaging apps.
> 
> 
> 
> We use better_player which provides Cache feature and we also added to it preCache. If preCache is enabled, the controller will cache video for only X ms (we set X = 350ms). This is enough to let video playback starts with no delay and at same time save cost of bandwidth. Lets consider a social media app caching tens of videos at once, that would be pricely, inefficient and unperformant. So preCache helps much here, we preCache them all and cache only the ones which get played.
> 
> 
> 
> Wish flutter team gets to enable both Cache/preCache this year. Although we have a solution, we think video playback should not rely on external plugins.

Very good points here. I have not used the better player package since it's not maintained so well. Do you have a fork that we can look at to see how your doing the precaching since you mentioned doing it on own?

**Comment by stuartmorgan** [comment]
Please see https://github.com/flutter/flutter/wiki/Issue-hygiene#comments-providing-workarounds; if there's going to be in depth discussions of alternatively, please link to a discussion elsewhere rather than having it here. This issue needs to remain focused on `video_player` itself.

**Comment by seljad** [comment]
any updates for cache videos? it's 2023!

**Comment by riveraj33** [comment]
> any updates for cache videos? it's 2023!

You can use flutter cache manager to download the videos but the official plug-in does not support precaching. Precaching is critical for social apps based on videos. The only solution is using better player but it hasn't been updated for over a year which is no good.

**Comment by seljad** [comment]
> > any updates for cache videos? it's 2023!
> 
> You can use flutter cache manager to download the videos but the official plug-in does not support precaching. Precaching is critical for social apps based on videos. The only solution is using better player but it hasn't been updated for over a year which is no good.

I have an application like Instagram. So I need to cache videos to users do not have to download videos every time. If I want to cache videos with cache manager, the user must wait for the whole video to download and then can play that. That's not good!

**Comment by riveraj33** [comment]
> > > any updates for cache videos? it's 2023!
> 
> > 
> 
> > You can use flutter cache manager to download the videos but the official plug-in does not support precaching. Precaching is critical for social apps based on videos. The only solution is using better player but it hasn't been updated for over a year which is no good.
> 
> 
> 
> I have an application like Instagram. So I need to cache videos to users do not have to download videos every time. If I want to cache videos with cache manager, the user must wait for the whole video to download and then can play that. That's not good!

How about using the HLS for videos. I'm not 100% sure but using HLS should only download the first segment and start to play before the entire video is downloaded first.

**Comment by seljad** [comment]
> > > > any updates for cache videos? it's 2023!
> > 
> > 
> > > 
> > 
> > 
> > > You can use flutter cache manager to download the videos but the official plug-in does not support precaching. Precaching is critical for social apps based on videos. The only solution is using better player but it hasn't been updated for over a year which is no good.
> > 
> > 
> > I have an application like Instagram. So I need to cache videos to users do not have to download videos every time. If I want to cache videos with cache manager, the user must wait for the whole video to download and then can play that. That's not good!
> 
> How about using the HLS for videos. I'm not 100% sure but using HLS should only download the first segment and start to play before the entire video is downloaded first.

video_player can play videos immediately. But if the user watches a video, navigates to the next video, and then returns to the first video, video_player will download it again even though the first video has been played and downloaded once. I just wanna videos cached when users see that.

**Comment by stuartmorgan** [comment]
@seljad @riveraj33 Please read my comment above.

**Comment by kishangohel** [comment]
When can we expect this to available?

**Comment by stuartmorgan** [comment]
Assigning myself to coordinate tracking of the in-progress PR.

**Comment by ciriousjoker** [comment]
@stuartmorgan thanks a lot! The cached_video_player package isn't being maintained anymore since the dev has no time. Supporting this in the official video player would be huge!

**Comment by kishangohel** [comment]
It will be better if we have buffer control too with caching @stuartmorgan

**Comment by kishangohel** [comment]
@stuartmorgan there is no activity or development going on [https://github.com/flutter/packages/pull/4526](PR)

**Comment by chorauoc** [comment]
Hi,

Any update on this request, we would like to have this feature 

Thanks

**Comment by kishangohel** [comment]
@stuartmorgan any update on this feature?
Is it work in progress?

**Comment by cosin666** [comment]
Really need this feature.

**Comment by priyesh-techabbot** [comment]
+1

**Comment by dhaval-patel-ke** [comment]
Hello Team,
Any progress on this issue?

**Comment by pirzada-jahanzaib** [comment]
i tried to use this package but it is giving this error

Because [AppName] depends on video_player from git which doesn't support null safety, version solving failed.

**Comment by ciriousjoker** [comment]
@stuartmorgan 
For perspective, our workout app uses lots of videos and absolutely requires caching. All the third party packages either have weird quirks with hls playback or aren't maintained anymore. They just slap a network caching layer on top and apparently that works stable without any maintenance, but I have the feeling that this can break at any point.

As for features we'd like to have:
- auto caching via a simple configurable bool on the video player
- manual caching key (video url might slightly change over time, key stays the same)
- some sort of offline capability: we dump a bunch of urls into the Videoplayer and have it force-cache the given urls until the user specifically removes them. Think of it like Spotify's offline playback. I understand that this will most likely be out of scope though.

**Comment by delfme** [comment]
Hello @stuartmorgan when feature will be reworked, you guys might also want to have a quick check at how it is implemented in `better_player`. 

It is not much code and has been tested for a while.
Android:
https://github.com/jhomlala/betterplayer/blob/master/android/src/main/kotlin/com/jhomlala/better_player/BetterPlayerCache.kt

IOS:
https://github.com/jhomlala/betterplayer/blob/master/ios/Classes/CacheManager.swift
https://github.com/jhomlala/betterplayer/blob/master/ios/Classes/CachingPlayerItem.swift

**Comment by stuartmorgan** [comment]
> you guys might also want to have a quick check at how it is implemented in `better_player`.

Anyone reading a different package's cache implementation code would be unable to write an implementation for `video_player` that is consistent with our CLA, so that is not something we would do, nor should anyone else interested in working on this feature.

**Comment by windows7lake** [comment]
https://pub.dev/packages/flutter_video_caching

**Comment by b3nni97** [comment]
@windows7lake 

I was wondering if you could provide more details about how the package handles video precaching, especially for .m3u8 streams.
When precaching an HLS video, does the package download and cache the entire video file, or just the first segment?

Also, when playing videos using the video_player package together with your caching package, does the video get downloaded twice, or is it downloaded only once in total?

**Comment by delfme** [comment]
@windows7lake I’ll take a look next week. In our app, we’ve implemented a native caching mechanism that allows us to pre-cache 1 second of data at 720p resolution, or a 200ms duration (the approach differs slightly between Android and iOS). @b3nni97 I believe this package does something similar, pre-caching isn’t about downloading the entire file, but enabling instant playback.

**Comment by b3nni97** [comment]
@delfme Yes, that's exactly why I followed up. So far, all the caching packages I've come across download the entire video, which is quite inefficient, especially if the user ends up not watching the video at all.
Would you be able to share your caching mechanism? We've been looking for a suitable solution for a while now, and since it seems like this issue might remain open for some time, we're exploring alternatives.

**Comment by windows7lake** [comment]
@delfme @b3nni97 You can try running the demo and test on it. As the documentation describes, you can specify the number of segments to cache using `VideoCaching.precache(url, segmentcount).` VideoCaching downloads the clips in a new thread and loads them into memory. If it is in m3u8 format, the first few ts files will also be downloaded. If it is mp4, the mp4 will be cut into 2mb size and the first few fragments will be downloaded.
I have been test and use it in my project. However, it is not perfect yet and needs to continue to be optimized and new features added.

**Comment by windows7lake** [comment]
> [@windows7lake](https://github.com/windows7lake)
> 
> I was wondering if you could provide more details about how the package handles video precaching, especially for .m3u8 streams. When precaching an HLS video, does the package download and cache the entire video file, or just the first segment?
> 
> Also, when playing videos using the video_player package together with your caching package, does the video get downloaded twice, or is it downloaded only once in total?

It will download it once and then get the segment from memory or storage file, if both failed, than download and save it.

**Comment by windows7lake** [comment]
@kishangohel You can try the demo, it will cached when you play the video.
